### ISSUE TITLE
Retry analyzer not calling BeforeMethod & AfterMethods if firstTimeOnly/lastTimeOnly are mentioned for the test

### ISSUE THREAD

**Issue by narulacool** [testng-team/testng#792]
Consider the case below:

```
@BeforeMethod(firstTimeOnly=true)
public void StartDriver()
{
   System.out.println("Start Driver");
}

@AfterMethod(lastTimeOnly=true)
public void stopDriver()
{
   System.out.println("Stop Driver");
}

@Test
public void TestRetry()
{
   Assert.assertTrue(false,"TestRetry Failed");
}
```

If I use TestNG's <i>IRetryAnalyzer</i> to rerun the above case, the StartDriver & stopDriver are called for the actual first run, while none of the two is called for the Second(retry) time.

```
[testng] Start Driver
[testng] [Status]:TestRetry2_FAIL
[testng] Stop Driver
[testng] Going to retry test case: TestRetry()
[testng] [Status]:TestRetry2_FAIL
```

I am using TestNG 6.9.4 & this works fine(i.e start & stop are called for both first time & for the retry execution) if firstTimeOnly & lastTimeOnly are not mentioned.

Any help would be deeply appreciated.

**Comment by narulacool** [comment]
ping..

**Comment by prvn133t** [comment]
I guess this is working as designed. Looking at the code of `public boolean firstTimeOnly() default false;`

I see 
> If true and the @Test method about to be run has an invocationCount > 1, this
BeforeMethod will only be invoked once (before the first test invocation).

So the `invocationCount` is already > 1 during the retry. So this is working as expected.

**Comment by rschmitt** [comment]
It does, however, look like the `@AfterMethod` is being called prematurely.

**Comment by prvn133t** [comment]
@rschmitt  : Do you have a test? I can debug the code.

**Comment by rschmitt** [comment]
Says who? Look at his console output again:

```
[testng] Start Driver
[testng] [Status]:TestRetry2_FAIL
[testng] Stop Driver
[testng] Going to retry test case: TestRetry()
[testng] [Status]:TestRetry2_FAIL
```

It's an invariant violation for the `@AfterMethod` here to run _between_ tests, isn't it?

**Comment by prvn133t** [comment]
@rschmitt  : I totally agree what you're saying. All I'm saying is, this defect context doesn't reflect actual bug in testng. 

There should be a new defect for "Retry analyzer not calling BeforeMethod & AfterMethods if it's specified in a `Test Class`"

**Comment by juherr** [comment]
Triage verdict: **still relevant**, although the expected retry semantics need to be made explicit before changing production code.

Current retry execution still interacts inconsistently with `firstTimeOnly` / `lastTimeOnly`. A retry re-enters `TestInvoker.invokeMethod(...)`, but the first attempt increments `ITestNGMethod.currentInvocationCount` before the retry is launched. `TestNgMethodUtils.doesSetupMethodPassFirstTimeFilter()` therefore suppresses a `firstTimeOnly` `@BeforeMethod` on the retry. Conversely, `doesTeardownMethodPassLastTimeFilter()` relies on `hasMoreInvocation()`, which tracks invocation-count/data-provider iterations, not whether `IRetryAnalyzer` has requested another attempt; a `lastTimeOnly` teardown can therefore become eligible before the retry lifecycle is actually over.

The current `FirstAndLastTimeTest` has good coverage for invocation counts, data providers, and the recent GITHUB-426 parallel barrier fix, but no retry-analyzer case. Recent fixes such as #2961/#3358 also changed this area without defining retry semantics. I found no #792 / `GITHUB-792` source/test/CHANGES entry, no newer canonical issue for this exact interaction, and no open or merged PR implementing it.

I did **not** execute the historical reproducer during this triage. The interaction is established directly by the current retry path and time-only configuration filters, but the original 2015 expectation ('run both configurations for every retry') is not the only coherent contract. Another reasonable model is that a retry is another attempt of the same invocation, in which case `firstTimeOnly` should run once before the first attempt and `lastTimeOnly` once after the final attempt.

Next step: encode the desired contract before a fix. The smallest regression should use a retry analyzer that retries exactly once, a deterministic first-fail/second-pass test, and `@BeforeMethod(firstTimeOnly=true)` / `@AfterMethod(lastTimeOnly=true)`, recording the full invocation order. Decide whether the expected order is `before, test-attempt-1, test-attempt-2, after` (retry stays inside one invocation) or a configuration pair around each retry attempt; then make that test fail on current code before modifying the retry/configuration lifecycle.
